walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,76 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+          
+      // take out events where all people that attend are not in request
+      List<Event> new_events = events.stream()
+        .filter(event -> !(Collections.disjoint(event.getAttendees(), request.getAttendees())))
+        .collect(Collectors.toList());
+
+      // sort events in ascending time
+      Collections.sort(new_events, Comparator.comparing(Event::getWhen,TimeRange.ORDER_BY_START));
+      
+      int tempStart;
+      int tempEnd;  
+      int temp = 0;  // this will keep track of the end time of the event that came before
+
+      int duration = (int)request.getDuration();
+
+      List<TimeRange> timeRanges = new ArrayList<TimeRange>();
+      
+      for ( int i = 0; i < new_events.size() ; i++ ){
+          tempStart = new_events.get(i).getWhen().start();
+          tempEnd = new_events.get(i).getWhen().end();
+
+          if (i == 0){
+            if ( duration <= tempStart){                                                  // the time before the first event   //xxx|--1--|---//
+              TimeRange first =  TimeRange.fromStartDuration(0,tempStart);
+               timeRanges.add(first);
```

Comment:
This case actually looks a lot like the next one, except here you're using `0` instead of `temp`. In this scenario (`i==0`), `temp` has to be `0`. These two could probably be unified and reduce some complexity.